### ISSUE TITLE
k256: get the correct signature from digest for eth

### ISSUE THREAD

**Issue by xphoniex** [RustCrypto/elliptic-curves#754]
I'm using `ethereum/go-ethereum` to generate digest and then a signature for a tx. I've tried to re-produce the same signature using `k256` yet I'm getting a different value.

Here's the setup:
```
private key:
0100000000000000000000000000000000000000000000000000000000000000

digest:
52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3

expecting signature:
71f498f5d9a05e83599a65e31febd4fb07451127031cf069c05c55ce586f248b0247be206cff797fe03223673560c431cf2f94a02ebbb09fda8e4a7b03c2f35301
```

I've tried `sha3::keccak256`:

```rust
fn main() {
    use hex_literal::hex;
    use k256::ecdsa::signature::DigestSigner;
    use k256::ecdsa::{Signature, SigningKey};

    use sha3::{Digest, Keccak256};

    let signing_key = SigningKey::from_bytes(&hex!(
        "0100000000000000000000000000000000000000000000000000000000000000"
    ))
    .unwrap();

    let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
    let sig: Signature = signing_key.sign_digest(Keccak256::new_with_prefix(msg));

    println!("{sig}");
}

// 3B90D82888D0E5A8428FC8F0956E2368150E292074082128611CEC323D6D37447DB3E5D98E1EBDB20250AC88701EF92C1596DBD30F0DDF0E1160E73E9025CA4F
```

and `sha2::Sha256`:

```rust
  fn main() {
      use hex_literal::hex;
      use k256::ecdsa::signature::DigestSigner;
      use k256::ecdsa::{Signature, SigningKey};
      use k256::sha2::{Digest, Sha256};

      let signing_key = SigningKey::from_bytes(&hex!(
          "0100000000000000000000000000000000000000000000000000000000000000"
      ))
      .unwrap();

      let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
      let sig: Signature = signing_key.sign_digest(Sha256::new_with_prefix(msg));

      println!("{sig}");
  }

// EED7F17FE57863ADA16A0BF1CB143D3E42E1E9A632C0CA8D6F9A8F970A85900200E32A25EA132ECA4033F324DC5EDBE7BD2F5E816B5654CD0A452CB96A2FA335
```

what's the correct way?

**Comment by tarcieri** [comment]
What you're doing looks correct. You should be using Keccak256 for Ethereum compatibility.

We test against the following vectors:

https://github.com/RustCrypto/elliptic-curves/pull/628/files

Those are in turn generated via web3.js:

https://web3js.readthedocs.io/en/v1.2.0/web3-eth-accounts.html#eth-accounts-signtransaction

I don't know offhand if web3.js and `ethereum/go-ethereum` produce identical signatures.

Note that what actually matters is that the other libraries can *verify* the signature. Producing identical signatures requires both libraries implement RFC6979 the same way (or even use RFC6979 at all), which may not be the case.

We implement RFC6979 using SHA-256, as was done with web3.js.

**Comment by tarcieri** [comment]
You might want to check to see if `ethereum/go-ethereum` is RLP-encoding the input message or applying something like EIP-155 or EIP-712 first.

We don't implement those because they're Ethereum-specific protocols. If you're looking for a Rust implementation of those, check out ethers-rs (which is built on `k256`):

https://github.com/gakonst/ethers-rs

**Comment by xphoniex** [comment]
> I don't know offhand if web3.js and ethereum/go-ethereum produce identical signatures.

I can get identical signatures if I try to re-produce a tx in go-ethereum which has been signed using metamask, which I suppose uses web3.js under the hood.

> You might want to check to see if ethereum/go-ethereum is RLP-encoding the input message or applying something like EIP-155 or EIP-712 first.

I'm calling:
```go
// main.go
sig, err := crypto.Sign(h[:], prv) // h = 52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3
```

which calls:
```go
// ethereum/go-ethereum@v1.10.19/crypto/signature_cgo.go
func Sign(digestHash []byte, prv *ecdsa.PrivateKey) (sig []byte, err error) {
    if len(digestHash) != DigestLength {
        return nil, fmt.Errorf("hash is required to be exactly %d bytes (%d)", DigestLength, len(digestHash))
    }
    seckey := math.PaddedBigBytes(prv.D, prv.Params().BitSize/8) // printing shows it remains the same
    defer zeroBytes(seckey)
    return secp256k1.Sign(digestHash, seckey)
}
```

which finally calls:
```go
// ethereum/go-ethereum@v1.10.19/crypto/secp256k1/secp256.go
func Sign(msg []byte, seckey []byte) ([]byte, error) {
    if len(msg) != 32 {
        return nil, ErrInvalidMsgLen
    }
    if len(seckey) != 32 {
        return nil, ErrInvalidKey
    }
    seckeydata := (*C.uchar)(unsafe.Pointer(&seckey[0]))
    if C.secp256k1_ec_seckey_verify(context, seckeydata) != 1 {
        return nil, ErrInvalidKey
    }

    var (
        msgdata   = (*C.uchar)(unsafe.Pointer(&msg[0]))
        noncefunc = C.secp256k1_nonce_function_rfc6979
        sigstruct C.secp256k1_ecdsa_recoverable_signature
    )
    if C.secp256k1_ecdsa_sign_recoverable(context, &sigstruct, msgdata, seckeydata, noncefunc, nil) == 0 {
        return nil, ErrSignFailed
    }

    var (
        sig     = make([]byte, 65)
        sigdata = (*C.uchar)(unsafe.Pointer(&sig[0]))
        recid   C.int
    )
    C.secp256k1_ecdsa_recoverable_signature_serialize_compact(context, sigdata, &recid, &sigstruct)
    sig[64] = byte(recid) // add back recid to get 65 bytes sig
    return sig, nil
}
```

I'm looking at [`secp256k1_ecdsa_sign_recoverable`](https://github.com/bitcoin-core/secp256k1/blob/aa5d34a8fe99b1f69306be20819f337dbd3283db/src/modules/recovery/main_impl.h#L123) (and [`secp256k1_ecdsa_sign_inner`](https://github.com/bitcoin-core/secp256k1/blob/5fbff5d348f4108dbf3111ef29840ed968054f6b/src/secp256k1.c#L502)) and they don't seem to be modifying the msg.

**Comment by xphoniex** [comment]
> Note that what actually matters is that the other libraries can _verify_ the signature. Producing identical signatures requires both libraries implement RFC6979 the same way (or even use RFC6979 at all), which may not be the case.

I can only verify `k256` signed case, not even `geth` one :confused: 

```rust
      #[test]
      fn test_sig_k256_signed() {
          let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
          let digest = Keccak256::new_with_prefix(msg);

          let sig = hex!("3B90D82888D0E5A8428FC8F0956E2368150E292074082128611CEC323D6D37447DB3E5D98E1EBDB20250AC88701EF92C1596DBD30F0DDF0E1160E73E9025CA4F");
          let sig = Signature::try_from(&sig[..]).unwrap();

          let recid = RecoveryId::new(false, false);
          let pk = VerifyingKey::recover_from_digest(digest, &sig, recid).unwrap();

          assert_eq!(VK, EncodedPoint::from(&pk).as_bytes(),);
      }

      #[test]
      fn test_sig_uECC_signed() {
          let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
          let digest = Keccak256::new_with_prefix(msg);

          let sig = hex!("f8f4a8dd90f010ed7b5348054ce43dbb1e71839143b1c1f13942c2289cec73c251a673eb714f337e3039619fb9b7e81af0e8b1177f811de0fd42b5dcefe4469c");
          let sig = Signature::try_from(&sig[..]).unwrap();

          let recid = RecoveryId::new(false, false);
          let pk = VerifyingKey::recover_from_digest(digest, &sig, recid).unwrap();

          assert_eq!(VK, EncodedPoint::from(&pk).as_bytes(),);
      }

      #[test]
      fn test_sig_geth_signed() {
          let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
          let digest = Keccak256::new_with_prefix(msg);

          let sig = hex!("71f498f5d9a05e83599a65e31febd4fb07451127031cf069c05c55ce586f248b0247be206cff797fe03223673560c431cf2f94a02ebbb09fda8e4a7b03c2f353");
          let sig = Signature::try_from(&sig[..]).unwrap();

          let recid = RecoveryId::new(false, false);
          let pk = VerifyingKey::recover_from_digest(digest, &sig, recid).unwrap();

          assert_eq!(VK, EncodedPoint::from(&pk).as_bytes(),);
      }
```

```rust
running 3 tests
test tests::test_sig_k256_signed ... ok
test tests::test_sig_geth_signed ... FAILED
test tests::test_sig_uECC_signed ... FAILED

failures:

---- tests::test_sig_geth_signed stdout ----
thread 'tests::test_sig_geth_signed' panicked at 'assertion failed: `(left == right)`
  left: `[2, 140, 40, 169, 123, 248, 41, 139, 192, 210, 61, 140, 116, 148, 82, 163, 46, 105, 75, 101, 227, 10, 148, 114, 163, 149, 74, 179, 15, 229, 50, 76, 170]`,
 right: `[3, 55, 123, 73, 242, 220, 40, 177, 77, 157, 165, 125, 173, 57, 142, 248, 41, 208, 36, 8, 13, 146, 157, 72, 3, 110, 51, 47, 62, 252, 185, 124, 189]`', src/main.rs:73:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::test_sig_uECC_signed stdout ----
thread 'tests::test_sig_uECC_signed' panicked at 'assertion failed: `(left == right)`
  left: `[2, 140, 40, 169, 123, 248, 41, 139, 192, 210, 61, 140, 116, 148, 82, 163, 46, 105, 75, 101, 227, 10, 148, 114, 163, 149, 74, 179, 15, 229, 50, 76, 170]`,
 right: `[2, 90, 141, 225, 227, 215, 48, 66, 22, 125, 203, 154, 157, 78, 32, 56, 21, 9, 83, 111, 44, 233, 54, 37, 205, 112, 124, 36, 156, 167, 192, 147, 216]`', src/main.rs:59:9


failures:
    tests::test_sig_geth_signed
    tests::test_sig_uECC_signed

test result: FAILED. 1 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

**Comment by tarcieri** [comment]
@xphoniex is there some reason you're trying to integrate with `k256` directly rather than using something like ethers-rs which already has well-tested handling of Ethereum signatures and message formats?

**Comment by tarcieri** [comment]
Also what would really be helpful here is to test against the `secp256k1` crate.

If you can demonstrate a mismatch behavior mismatch with the `secp256k1` crate, that would be a legitimate bug.

**Comment by xphoniex** [comment]
> @xphoniex is there some reason you're trying to integrate with `k256` directly rather than using something like ethers-rs which already has well-tested handling of Ethereum signatures and message formats?
> 

ethers-rs simply calls [`self.signer.sign_digest(Sha256Proxy::from(hash))`](https://github.com/gakonst/ethers-rs/blob/1b08a94a858186f825b77517f14bc865cd63a597/ethers-signers/src/wallet/mod.rs#L154) and I thought it'd be straightforward, that's why I attempted to use `k256` directly.


I copied the [`Sha256Proxy`](https://github.com/gakonst/ethers-rs/blob/1b08a94a858186f825b77517f14bc865cd63a597/ethers-signers/src/wallet/hash.rs#L11) type and it's impls over and now I'm getting the same sig as go-ethereum. Not sure why `Sha256::new_with_prefix` doesn't work... does it hash what it receives? And any idea why `test_sig_geth_signed` fails?

**Comment by tarcieri** [comment]
No idea, it looks like it should work.

Again, the best way to confirm your test vector is even valid is to test it against the `secp256k1` crate.

**Comment by xphoniex** [comment]
Btw, this too breaks once you update to `k256` to `"0.12"`, to get it compiled, you'd have to implement another trait:
```rust
  impl<D: Digest> BlockSizeUser for ProxyDigest<D> {
      type BlockSize = D::OutputSize;
  }
```

and with no other changes, you get a different signature. ether-rs is also using [`"0.11"`](https://github.com/gakonst/ethers-rs/blob/1b08a94a858186f825b77517f14bc865cd63a597/ethers-core/Cargo.toml#L23).

**Comment by tarcieri** [comment]
Can you post a full reproduction?

**Comment by xphoniex** [comment]
**main.rs**

```rust
use hex_literal::hex;
use k256::ecdsa::signature::DigestSigner;
use k256::ecdsa::SigningKey;
use k256::ecdsa::Signature;

use k256::sha2::{Digest, Sha256};
//use sha3::{Digest, Keccak256};

use k256::ecdsa::signature::digest::generic_array::GenericArray;
use k256::ecdsa::signature::digest::{
    core_api::BlockSizeUser, FixedOutput, FixedOutputReset, HashMarker, Output, OutputSizeUser,
    Reset, Update,
};

#[derive(Clone)]
pub enum ProxyDigest<D: Digest> {
    Proxy(Output<D>),
    Digest(D),
}

impl<D: Digest> BlockSizeUser for ProxyDigest<D> {
    type BlockSize = D::OutputSize;

    fn block_size() -> usize {
        64
    }
}

impl<D: Digest> Default for ProxyDigest<D> {
    fn default() -> Self {
        ProxyDigest::Digest(D::new())
    }
}

impl<D: Digest> Update for ProxyDigest<D> {
    // we update only if we are digest
    fn update(&mut self, data: &[u8]) {
        match self {
            ProxyDigest::Digest(ref mut d) => {
                d.update(data);
            }
            ProxyDigest::Proxy(..) => {
                unreachable!("can not update if we are proxy");
            }
        }
    }
}

impl<D: Digest> HashMarker for ProxyDigest<D> {}

impl<D: Digest> Reset for ProxyDigest<D> {
    // make new one
    fn reset(&mut self) {
        *self = Self::default();
    }
}

impl<D: Digest> OutputSizeUser for ProxyDigest<D> {
    // we default to the output of the original digest
    type OutputSize = <D as OutputSizeUser>::OutputSize;
}

impl<D: Digest> FixedOutput for ProxyDigest<D> {
    fn finalize_into(self, out: &mut GenericArray<u8, Self::OutputSize>) {
        match self {
            ProxyDigest::Digest(d) => {
                *out = d.finalize();
            }
            ProxyDigest::Proxy(p) => {
                *out = p;
            }
        }
    }
}

impl<D: Digest> FixedOutputReset for ProxyDigest<D> {
    fn finalize_into_reset(&mut self, out: &mut Output<Self>) {
        let s = std::mem::take(self);
        Digest::finalize_into(s, out)
    }
}

fn main() {
    let signing_key = SigningKey::from_bytes(&hex!(
        "0100000000000000000000000000000000000000000000000000000000000000"
    ))
    .unwrap();

    let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
    let msg: ProxyDigest<Sha256> = ProxyDigest::Proxy(*GenericArray::from_slice(&msg));

    let sig: Signature = signing_key.sign_digest(msg);
    //let sig: Signature = signing_key.sign_digest(Sha256::new_with_prefix(msg));
    //let sig: Signature = signing_key.sign_digest(Keccak255::new_with_prefix(msg));

    println!("{sig}");
}

#[cfg(test)]
mod tests {
    use hex_literal::hex;
    use k256::ecdsa::{RecoveryId, Signature, VerifyingKey};
    use k256::EncodedPoint;
    use sha3::{Digest, Keccak256};

    const VK: [u8; 33] = hex!("028c28a97bf8298bc0d23d8c749452a32e694b65e30a9472a3954ab30fe5324caa");

    #[test]
    fn test_sig_k256_signed() {
        let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
        let digest = Keccak256::new_with_prefix(msg);

        let sig = hex!("3B90D82888D0E5A8428FC8F0956E2368150E292074082128611CEC323D6D37447DB3E5D98E1EBDB20250AC88701EF92C1596DBD30F0DDF0E1160E73E9025CA4F");
        let sig = Signature::try_from(&sig[..]).unwrap();

        let recid = RecoveryId::new(false, false);
        let pk = VerifyingKey::recover_from_digest(digest, &sig, recid).unwrap();

        assert_eq!(VK, EncodedPoint::from(&pk).as_bytes(),);
    }

    #[test]
    fn test_sig_uECC_signed() {
        let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
        let digest = Keccak256::new_with_prefix(msg);

        let sig = hex!("f8f4a8dd90f010ed7b5348054ce43dbb1e71839143b1c1f13942c2289cec73c251a673eb714f337e3039619fb9b7e81af0e8b1177f811de0fd42b5dcefe4469c");
        let sig = Signature::try_from(&sig[..]).unwrap();

        let recid = RecoveryId::new(false, false);
        let pk = VerifyingKey::recover_from_digest(digest, &sig, recid).unwrap();

        assert_eq!(VK, EncodedPoint::from(&pk).as_bytes(),);
    }

    #[test]
    fn test_sig_geth_signed() {
        let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
        let digest = Keccak256::new_with_prefix(msg);

        let sig = hex!("71f498f5d9a05e83599a65e31febd4fb07451127031cf069c05c55ce586f248b0247be206cff797fe03223673560c431cf2f94a02ebbb09fda8e4a7b03c2f353");
        let sig = Signature::try_from(&sig[..]).unwrap();

        let recid = RecoveryId::new(false, false);
        let pk = VerifyingKey::recover_from_digest(digest, &sig, recid).unwrap();

        assert_eq!(VK, EncodedPoint::from(&pk).as_bytes(),);
    }
}
```

**Cargo.toml**
```toml
[dependencies]
#k256 = { version = "0.12.0", features = ["ecdsa", "ecdsa-core", "digest", "sha256", "schnorr"] }
k256 = { version = "0.11.2", features = ["ecdsa", "ecdsa-core", "digest", "sha256", "schnorr"] }
hex-literal = "0.3.4"
sha3 = "0.10.1"
hex = "0.4"
```

**0.11.2 output:** (same as go-ethereum)

```
71F498F5D9A05E83599A65E31FEBD4FB07451127031CF069C05C55CE586F248B0247BE206CFF797FE03223673560C431CF2F94A02EBBB09FDA8E4A7B03C2F353
```

**0.12.0 output:**

```
E5DDAF165DBC7FF272EA9869B186228CF2FBC461A8A140B266AD82F41BDC14C868AB5B9FAD8032F0363F3D24CA03FCEAFF73E402CB15C82C5D9FCCAB440588B4
```

**Comment by tarcieri** [comment]
Still no idea what's happening there. The code is quite convoluted unfortunately.

Here is a complete code example which tests against the ethers-rs vector `k256` uses for testing, which was in turn generated by web3.js:

https://docs.rs/k256/latest/k256/ecdsa/index.html#computing-a-signature-with-a-recoveryid

I am not sure under what circumstances that would work, but your other examples would not.

**Comment by tarcieri** [comment]
This PR updates the example to use the `sign_digest_recoverable` method:

https://github.com/RustCrypto/elliptic-curves/pull/755/files

**Comment by tarcieri** [comment]
Here's a complete end-to-end example which tests against the web3.js-generated test vectors, ensuring the expected signature is produced along with the correct `recid`, and then further using that to recover the public key and ensuring the signature verifies:

https://github.com/RustCrypto/elliptic-curves/pull/756/files

**Comment by xphoniex** [comment]
> Still no idea what's happening there. The code is quite convoluted unfortunately.

All those borrowed impls (from ethers-rs)  are required just to pass **actual** digest into `sign_digest` fn.

**Comment by xphoniex** [comment]
Ok, I figured it out. 

- `sign_digest` in `v0.12.0` is not the same as previous versions, instead you should use `try_sign_prehashed_rfc6979` (might wanna revert https://github.com/RustCrypto/elliptic-curves/pull/755 , unless you add a fix), as you can't still use `sign_digest_recoverable` for raw digest

- The issue is with `Keccak256::new_with_prefix`. I was expecting this fn to take **actual** keccak256 digest, not bytes that will be turned into keccack256 digest. Test cases in-tree work with signing _messages_, not digests. The issue is, if you try with raw `&[u8]`, it won't compile because of needed traits. The only thing that works is `GenericArray`:

This does **not** work:
```rust
let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
let digest = Keccak256::digest(msg); // <---
let (sig, recid) = signing_key
    .as_nonzero_scalar()
    .try_sign_prehashed_rfc6979::<k256::sha2::Sha256>(digest, b"")
    .unwrap();       
```

Neither this:
```rust
let msg = Keccak256::new_with_prefix(hex!(
    "52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3"
));
let (sig, recid) = signing_key.sign_digest_recoverable(msg).unwrap();
```

Only this works:
```rust
let msg = hex!("52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3");
let (sig, recid) = signing_key
    .as_nonzero_scalar()
    .try_sign_prehashed_rfc6979::<k256::sha2::Sha256>(*GenericArray::from_slice(&msg), b"")
    .unwrap();
    
// sig = 71f498f5d9a05e83599a65e31febd4fb07451127031cf069c05c55ce586f248b
//       0247be206cff797fe03223673560c431cf2f94a02ebbb09fda8e4a7b03c2f353
//       (same as web3.js & go-ethereum)
```

Here's `web3.js` code to re-produce this:
```javascript
const Web3 = require("web3");
const web3 = new Web3('ws://localhost:8545');

web3.eth.accounts.signTransaction({
    nonce: '0',
    chainID: '1',
    to: '0x0475F0d4a405A79b58f302BD22ECbdAF35B1759e',
    value: '1000000000',
    gas: '21000',
    maxPriorityFeePerGas: '1000000000',
    maxFeePerGas: '15000000000',
}, '0x0100000000000000000000000000000000000000000000000000000000000000')
.then(console.log);

// {
//  messageHash: '0x52fbb559a867af2f7613fd01b0a59e1363d3403830551c22f65622ccf93e7db3',
//  v: '0x1',
//  r: '0x71f498f5d9a05e83599a65e31febd4fb07451127031cf069c05c55ce586f248b',
//  s: '0x247be206cff797fe03223673560c431cf2f94a02ebbb09fda8e4a7b03c2f353',
//  rawTransaction:  '0x02f86f0180843b9aca0085037e11d600825208940475f0d4a405a79b58f302bd22ecbdaf35b1759e843b9aca0080c001a071f498f5d9a05e83599a65e31febd4fb07451127031cf069c05c55ce586f248ba00247be206cff797fe03223673560c431cf2f94a02ebbb09fda8e4a7b03c2f353',
//  transactionHash: '0xb87f16b34c90bd5e61a265f28768802283a40cf4c01eb6ac1f7a56d6d16b09d5'
// }
```

still I can't verify the sig, because unlike `try_sign_prehashed_rfc6979`, this doesn't compile:

```rust
let recid = recid.unwrap();
let recovered_key =
    VerifyingKey::recover_from_digest(*GenericArray::from_slice(&msg), &sig, recid).unwrap();
let expected_key = signing_key.verifying_key();

assert_eq!(recovered_key, *expected_key);
```

and can't find `recover_verifying_key_from_digest` method on sig either.

long story short: you need a method/type that accepts raw keccak256 hash, and implements all needed traits. 

p.s: also can I add test cases for digest signing in a PR?

**Comment by tarcieri** [comment]
> might wanna revert https://github.com/RustCrypto/elliptic-curves/pull/755

What? Why? It's doctested and working.

> The issue is with `Keccak256::new_with_prefix`. I was expecting this fn to take **actual** keccak256 digest, not bytes that will be turned into keccack256 digest.

The `*_digest` APIs work with types that impl the `Digest` trait, i.e. implementations of digest functions.

If you've already computed the digest output as bytes, you want to use the `*_prehash` APIs.

> still I can't verify the sig, because unlike `try_sign_prehashed_rfc6979`, this doesn't compile:

You're still trying to pass bytes of the digest output rather than a `Digest` instance.

Use [`PrehashVerifier::verify_prehash`](https://docs.rs/signature/latest/signature/hazmat/trait.PrehashVerifier.html#tymethod.verify_prehash).

> long story short: you need a method/type that accepts raw keccak256 hash,

That's all of the many `*_prehash` APIs.

> p.s: also can I add test cases for digest signing in a PR?

You can add test cases, but what would they demonstrate other than #755 and #756?

**Comment by tarcieri** [comment]
Going to close this. We can potentially document this a lot better, I think, but AFAICT that's the only thing actionable here.
